## AIVLE School [AI 미니프로젝트] '서울시 생활정보 기반 대중교통 수요 분석'<br>

---

# 2. 구 별 이동인구 분석
### 서울시 생활이동 정보를 기반으로 서울시 자치구 별 이동인구를 분석합니다

In [ ]:
# 필요 라이브러리를 설치합니다.
%pip install pandas seaborn

## 데이터 안내

[데이터 파일명]
* 2. seoul_moving.csv

[데이터 정보]
* 서울 시 구별 이동 2022년 4월 데이터

[데이터 컬럼 정보]
* 도착시간 : 0 -> "0:00~0:59"를 의미 
* 출발 자치구 : 자치구 코드
* 성별 : F(여성), M(남성)
* 나이 : 10 ~ 79세 (5세 단위), 0 ~ 10세, 80세 이상으로 구분
* 유형 : H(야간상주지), W(주간상주지), E(기타)
* 평균 이동 시간 : 동일 열 내 이동인구의 월 단위 평균 이동 시간 (10분단위)
* 이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만(0~3)의 경우 * 처리

---

# 2.1.데이터 불러오기

#### [2.1.1] 데이터 로딩
* Pandas 라이브러리를 활용해서 '2. seoul_moving.csv'파일을 'seoul_moving' 변수에 저장하세요.
    * 데이터 파일 로딩시 참고 사항 
        * 구분자(sep)는 ',' 입니다
        * UTF-8 인코더를 사용해 주세요
        * index column은 0번 열 입니다
    * 데이터를 불러올 때 오류가 발생한다면, 오류에 대해 직접 검색하여 해결해보세요.

In [12]:
import sys
import numpy as np
import pandas as pd

In [20]:
# 아래에 코드를 작성하고 결과를 확인합니다.
seoul_moving = pd.read_csv('2. seoul_moving.csv', encoding='utf-8', sep=',')

In [3]:
# 데이터 프레임의 Shape을 확인합니다.


---

# 2.2.기본 정보 확인 및 클렌징

+ 데이터 클렌징 : 결측치, 이상치 등을 제거하여 데이터 분석 결과가 왜곡 되는 문제를 방지하기 위한 정제 과정

#### [2.2.1] 기본 정보 확인하기
* 'seoul_moving' 데이터의 정보를 확인해보세요.
* 'describe', 'info', 'head', 'tail' 등 전부 활용해 보세요.

In [4]:
# 아래에 코드를 작성하고 결과를 확인합니다.


In [5]:
# 아래에 코드를 작성하고 결과를 확인합니다.


In [6]:
# 아래에 코드를 작성하고 결과를 확인합니다.


In [7]:
# 아래에 코드를 작성하고 결과를 확인합니다.


* [참고] 1번 열의 의미
    * " 2022년 4월 일요일에 11010(종로구)에서 11010(종로구)으로 이동하여 0시 ~ 0시 59분 사이에 도착했고, E(기타)에서 H(야간상주지(집))으로 이동한 0 ~ 10세 여성은 총 27.52명이며 이들의 평균 이동시간은 10분이다

#### [2.2.2] 결측치 처리 (NaN이 아닌 결측치 처리)

* '이동인구(합)' column은 이동 인구의 합산인데 숫자가 아닌 object 타입입니다. 

    - 왜 일까요? 

    - Tip! (이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만 즉, 값이 0~3일 경우 별표 * 처리)

    - 해당 값을 어떻게 처리하는 것이 좋을 지 생각해보아서 각자의 방법으로 변경하세요

    - '인구이동(합)' 컬럼의 데이터 타입을 수치형으로 변경하세요 

In [8]:
# 이동인구(합) 컬럼의 값 별로 개수를 세어봅니다.


In [21]:
# 별표(*)로 표현된 값을 대체합니다.
# 해당 값을 어떻게 처리하는 것이 좋을 지 생각해보아서 각자의 방법으로 변경하세요
#(ex: 최빈값, 평균값도 알 수 없으므로 0~3의 중위값인 1.5로 대체)
seoul_moving['이동인구(합)'] = seoul_moving['이동인구(합)'].replace('*', 1.5)

In [15]:
# 이동인구(합) 컬럼의 값 별로 개수를 다시 세어봅니다.


In [22]:
# 이제 이동인구(합) column의 데이터 타입을 바꿔주세요
seoul_moving['이동인구(합)']  = seoul_moving['이동인구(합)'].astype('float')

#### [2.2.3] 이동유형, 요일, 성별 변수의 value수 파악

In [12]:
# 이동유형 컬럼의 값 별로 개수를 세어봅니다.
# H(야간상주지), W(주간상주지), E(기타)
seoul_moving['이동유형'].value_counts()

EH    5757073
HE    5520005
EE    5085099
WH    3108026
WE    2809148
HW    2695189
EW    2304535
WW    1026134
HH     617845
Name: 이동유형, dtype: int64

In [13]:
# 요일 컬럼의 값 별로 개수를 세어봅니다.


* [참고] 2022년 4월의 요일 수는 다음과 같습니다.
    * 일 : 4, 월 : 4, 화 : 4, 수 : 4, 목 : 4, 금 : 5, 토 : 5

In [14]:
# 성별 컬럼의 값 별로 개수를 세어봅니다.



#### [2.2.4] 심야 이동이 가장 많은 요일 확인

In [17]:
# 아래에 코드를 작성하고 결과를 확인합니다.
# 심야시간은 0:00~ 4:59
# 아래에 코드를 작성하고 결과를 확인합니다.
seoul_moving = seoul_moving[seoul_moving['도착시간'] < 5]


Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
0                 0  202204  일     0      11010      11010  F   0   EE   
1                 1  202204  일     0      11010      11010  F   0   EH   
2                 2  202204  일     0      11010      11010  F   0   HE   
3                 3  202204  일     0      11010      11010  F  10   HW   
4                 4  202204  일     0      11010      11010  F  10   EH   
...             ...     ... ..   ...        ...        ... ..  ..  ...   
2292250      320268  202204  토     4      38000      11240  M  60   HE   
2292251      320269  202204  토     4      38000      11250  F  55   EH   
2292252      320270  202204  토     4      39000      11010  M  45   HE   
2292253      320271  202204  토     4      39000      11190  M  50   EE   
2292254      320272  202204  토     4      39000      11230  M  55   EE   

         평균 이동 시간(분)  이동인구(합)  
0                 10    28.02  
1                 10    27.52  
2                 10    27.83  
3                 10     6.05  
4                 90     5.95  
...              ...      ...  
2292250           90     3.56  
2292251          100     3.25  
2292252          410     3.15  
2292253          140     3.28  
2292254          310     3.30  

[2292255 rows x 11 columns]

In [19]:
seoul_moving.groupby('요일')[['이동인구(합)']].sum()

이동인구(합)
요일            
금   4206523.57
목   3311174.58
수   3203277.45
월   2511032.17
일   3550710.46
토   5233200.14
화   3035154.60

#### [2.2.5] 서울 지역의 데이터만 추출

* 서울시의 데이터만 필요하므로 출발과 도착이 모두 서울인 데이터만 추출하세요.
* 아래 사이트에서 서울시 데이터만 추출하기 위해 필요한 정보를 탐색해보세요. (ex. 자치구 코드 정보)
https://data.seoul.go.kr/dataVisual/seoul/seoulLivingMigration.do

In [23]:
# 출발 시군구 코드, 도착 시군구 코드 활용하여 출발과 도착이 모두 서울인 데이터만 추출해보세요.
seoul_moving = seoul_moving[(seoul_moving['출발 시군구 코드'] > 11000) & (seoul_moving['출발 시군구 코드'] < 20000)]
seoul_moving = seoul_moving[(seoul_moving['도착 시군구 코드'] > 11000) & (seoul_moving['도착 시군구 코드'] < 20000)]
seoul_moving

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
0                  0  202204  일     0      11010      11010  F   0   EE   
1                  1  202204  일     0      11010      11010  F   0   EH   
2                  2  202204  일     0      11010      11010  F   0   HE   
3                  3  202204  일     0      11010      11010  F  10   HW   
4                  4  202204  일     0      11010      11010  F  10   EH   
...              ...     ... ..   ...        ...        ... ..  ..  ...   
28875001      920410  202204  토    23      11250      11250  M  80   WH   
28875002      920411  202204  토    23      11250      11250  M  80   EH   
28875003      920412  202204  토    23      11250      11250  M  80   EE   
28875004      920413  202204  토    23      11250      11250  M  80   HE   
28875005      920414  202204  토    23      11250      11250  M  80   WE   

          평균 이동 시간(분)  이동인구(합)  
0                  10    28.02  
1                  10    27.52  
2                  10    27.83  
3                  10     6.05  
4                  90     5.95  
...               ...      ...  
28875001           10    85.10  
28875002           10    28.36  
28875003           20    37.99  
28875004           10     9.42  
28875005           10     9.51  

[10655895 rows x 11 columns]

#### [2.2.6] 서울지역의 구 별로 groupby하여 데이터를 확인해보세요. (유동인구 기준)

* 유출이 제일 많은 구는?

* 유입이 제일 많은 구는?

* 유출이 제일 적은 구는?

* 유입이 제일 적은 구는?

* [참고] 서울 구별 코드
    * 11010	종로구
    * 11020	중구
    * 11030	용산구
    * 11040	성동구
    * 11050	광진구
    * 11060	동대문구
    * 11070	중랑구
    * 11080	성북구
    * 11090	강북구
    * 11100	도봉구
    * 11110	노원구
    * 11120	은평구
    * 11130	서대문구
    * 11140	마포구
    * 11150	양천구
    * 11160	강서구
    * 11170	구로구
    * 11180	금천구
    * 11190	영등포구
    * 11200	동작구
    * 11210	관악구
    * 11220	서초구
    * 11230	강남구
    * 11240	송파구
    * 11250	강동구

In [31]:
# 아래에 코드를 작성하고 결과를 확인합니다.
seoul_moving.groupby('출발 시군구 코드', as_index=False)[['이동인구(합)']].sum().sort_values('이동인구(합)')

출발 시군구 코드      이동인구(합)
17      11180   8467440.48
9       11100  11253816.27
8       11090  13084060.28
6       11070  14509927.80
16      11170  15038178.22
2       11030  15148400.31
14      11150  15668364.41
19      11200  16472620.89
12      11130  16952589.75
5       11060  16973103.47
11      11120  17191013.47
3       11040  18121271.09
4       11050  18675122.83
7       11080  19558088.65
20      11210  19706689.55
0       11010  19862431.70
1       11020  20296645.78
10      11110  21392595.19
24      11250  22035184.61
15      11160  22526485.94
13      11140  24517465.86
18      11190  25835980.43
21      11220  29698071.20
23      11240  36207933.02
22      11230  46220924.63

In [49]:
# 아래에 코드를 작성하고 결과를 확인합니다.



#### [2.2.7] 낮시간만 분석 대상에 넣기 위해 심야버스 제외

* 도착시간 0부터 4까지(0:00 ~ 4:59) 제거

In [17]:
# 아래에 코드를 작성하고 결과를 확인합니다.
seoul_moving = seoul_moving[seoul_moving['도착시간'] < 5]
seoul_moving

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
0                 0  202204  일     0      11010      11010  F   0   EE   
1                 1  202204  일     0      11010      11010  F   0   EH   
2                 2  202204  일     0      11010      11010  F   0   HE   
3                 3  202204  일     0      11010      11010  F  10   HW   
4                 4  202204  일     0      11010      11010  F  10   EH   
...             ...     ... ..   ...        ...        ... ..  ..  ...   
2279524      307542  202204  토     4      11250      11250  M  80   WH   
2279525      307543  202204  토     4      11250      11250  M  80   WE   
2279526      307544  202204  토     4      11250      11250  M  80   EE   
2279527      307545  202204  토     4      11250      11250  M  80   EH   
2279528      307546  202204  토     4      11250      11250  M  80   WW   

         평균 이동 시간(분)  이동인구(합)  
0                 10    28.02  
1                 10    27.52  
2                 10    27.83  
3                 10     6.05  
4                 90     5.95  
...              ...      ...  
2279524           10    28.34  
2279525           10    85.68  
2279526           10    38.07  
2279527           10    28.36  
2279528           10    19.04  

[1099957 rows x 11 columns]

In [18]:
# '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)', '이동인구(합)'의 평균을 도출하여 seoul_moving_mean에 저장
seoul_moving_mean = seoul_moving.groupby('도착 시군구 코드', as_index=False)[['평균 이동 시간(분)', '이동인구(합)']].mean()


In [19]:
# '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)', '이동인구(합)'의 합을 도출하여 seoul_moving_sum에 저장
seoul_moving_sum = seoul_moving.groupby('도착 시군구 코드', as_index=False)[['평균 이동 시간(분)', '이동인구(합)']].sum()


In [20]:
# 필요에 따라 column명을 데이터 의미에 맞게 적절히 변경하세요.
seoul_moving_sum.rename(columns={'평균 이동 시간(분)':'총 이동시간'}, inplace=True)
seoul_moving_sum.rename(columns={'이동인구(합)':'총 이동인구'}, inplace = True)
# seoul_moving_sum.rename(columns = {'정류소명' : '정류소개수'}, inplace = True)
seoul_moving_sum

도착 시군구 코드   총 이동시간      총 이동인구
0       11010   883220   424279.03
1       11020   979390   560222.82
2       11030   973740   540110.37
3       11040  1065070   626961.55
4       11050  1136230   767907.45
5       11060  1160130   699500.69
6       11070  1163080   611038.12
7       11080  1241700   767082.61
8       11090  1115550   508611.41
9       11100  1057250   422423.27
10      11110  1280950   759880.83
11      11120  1144790   673147.12
12      11130  1004640   606865.00
13      11140  1227470   899490.45
14      11150  1065410   589063.90
15      11160  1330770   931139.16
16      11170  1125610   548678.17
17      11180   828250   341318.73
18      11190  1260960   817252.81
19      11200  1120300   698454.56
20      11210  1319560   974644.26
21      11220  1198160   852112.22
22      11230  1466820  1467961.97
23      11240  1368090  1319285.45
24      11250  1166180   894897.06

In [23]:
# seoul_moving_mean, seoul_moving_sum을 합쳐보세요.
df_seoul_moving = pd.merge(seoul_moving_mean, seoul_moving_sum, how='inner')
df_seoul_moving


도착 시군구 코드  평균 이동 시간(분)    이동인구(합)   총 이동시간      총 이동인구
0       11010    22.489242  10.803326   883220   424279.03
1       11020    22.748473  13.012399   979390   560222.82
2       11030    23.358921  12.956637   973740   540110.37
3       11040    23.879422  14.056804  1065070   626961.55
4       11050    25.393452  17.161861  1136230   767907.45
5       11060    24.698857  14.892182  1160130   699500.69
6       11070    27.324797  14.355412  1163080   611038.12
7       11080    25.430602  15.710214  1241700   767082.61
8       11090    28.094543  12.809112  1115550   508611.41
9       11100    30.085083  12.020468  1057250   422423.27
10      11110    28.369101  16.829022  1280950   759880.83
11      11120    28.822226  16.947736  1144790   673147.12
12      11130    25.239040  15.245949  1004640   606865.00
13      11140    24.736407  18.126848  1227470   899490.45
14      11150    26.820310  14.828917  1065410   589063.90
15      11160    28.479070  19.926793  1330770   931139.16
16      11170    28.231296  13.761335  1125610   548678.17
17      11180    27.930465  11.510040   828250   341318.73
18      11190    24.103682  15.622067  1260960   817252.81
19      11200    25.224028  15.726000  1120300   698454.56
20      11210    27.471374  20.290716  1319560   974644.26
21      11220    24.118038  17.152363  1198160   852112.22
22      11230    24.880754  24.900125  1466820  1467961.97
23      11240    27.057137  26.091914  1368090  1319285.45
24      11250    29.785962  22.856995  1166180   894897.06

In [22]:
# 해당 데이터프레임을 'df_seoul_moving.csv' 파일로 저장하세요
df_seoul_moving.to_csv('df_seoul_moving_야간.csv')

target = '도착 시군구 코드'
var = '평균 이동 시간(분)'
data = df_seoul_moving
plt.figure(figsize=(12,4))


plt.hist(x=target, data = data, hue = var)
plt.grid()
plt.show()

---

# 2.3.데이터 분석하기
+ 데이터의 형태를 살펴보고 다양한 분석기법을 통해 모델링에 적합하도록 정제요소를 선별하세요.
  * 데이터들의 패턴 탐색
  * 변수들간의 관계 파악

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# 시각화 한글폰트 설정
# 맑은 고딕

plt.rc('font', family='Malgun Gothic')
sns.set(font="Malgun Gothic",#"NanumGothicCoding", 
        rc={"axes.unicode_minus":False}, # 마이너스 부호 깨짐 현상 해결
        style='darkgrid')

# scipy
import scipy.stats as spst

#### [2.3.1] 데이터 분포 알아보기
* 다양한 변수를 기준으로 그래프를 그려보고 인사이트를 도출해보세요.

In [ ]:
# 아래에 코드를 작성하고 결과를 확인합니다.


In [ ]:
# 아래에 코드를 작성하고 결과를 확인합니다.

In [ ]:
# 아래에 코드를 작성하고 결과를 확인합니다.

In [ ]:
# 위 차트를 통해 알게된 사실을 정리해봅시다.
# 1. 
# 2.
# 3.

---